Eigen/src/Core/ProductEvaluators.h

```diff
@@ -137,6 +137,7 @@ struct Assignment<DstXprType, Product<Lhs,Rhs,Options>, internal::assign_op<Scal
   typename enable_if<(Options==DefaultProduct || Options==AliasFreeProduct)>::type>
 {
   typedef Product<Lhs,Rhs,Options> SrcXprType;
+  EIGEN_DEVICE_FUNC
```

Comment:
Was this an upstream bug that this definition was missing? Does the change work w/ CUDA?

Reply:
I need to guard this with `#if defined(EIGEN_HIPCC)`

The EIGEN_DEVICE_FUNC marking is somewhat puzzling to me. It expands to "__host__ __device__" for both CUDA and HIP. My assumption is that the code compile correctly "as is" for CUDA. However for HIP there are places where we need to either add or remove EIGEN_DEVICE_FUNC marking from a functions (mostly via trial and error based on compiler errors when compiling the unit tests). This leads me to believe that those attributes somehow get handled differently by NVCC / HIPCC ... wonder what the differences are!